xla/service/gpu/gpu_compiler.cc

```diff
@@ -1570,7 +1570,20 @@ absl::Status GpuCompiler::OptimizeHloPostLayoutAssignment(
     return Compiler::TargetConfig{gpu_target_config_proto};
   }
   if (executor) {
-    return Compiler::TargetConfig{executor};
+    Compiler::TargetConfig target_config = Compiler::TargetConfig{executor};
+    int64_t device_memory_size =
+      target_config.device_description.device_memory_size();
+    // Checking for device_memory_size == -1 is how we detect that we are
+    // running on Nvidia's software simulator. When running on simulation,
+    // the config from StreamExecutor is inaccurate, so we must load the
+    // hard-coded config from a file.
+    if (device_memory_size == -1) {
+      return absl::FailedPreconditionError(
+        "When running on an NVIDIA simulation device, you must use "
```

Comment:
Could you specify in the error message that we've implied it to be an NVIDIA simulator from memory size being minus one? 

Additionally, taking a step back, is there a more principled way to do this? Maybe some CUDA function which returns a known sentinel value? Maybe we should pipe that through instead?

Reply:
> Could you specify in the error message that we've implied it to be an NVIDIA simulator from memory size being minus one?

I intentionally left that out because it's an obscure implementation detail that should not be relevant to the user.

> Additionally, taking a step back, is there a more principled way to do this? Maybe some CUDA function which returns a known sentinel value? Maybe we should pipe that through instead?

It could also be detected via the device name (available from the field device_description_str). That was my first implementation (before sending out the PR). But the name of the simulator is an internal Nvidia name. I asked inside Nvidia and I didn't get permission to open source the name.